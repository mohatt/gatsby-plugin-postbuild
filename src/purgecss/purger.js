import { Promise } from 'bluebird'
import { promises as fs } from 'fs'
import PurgeCSS from 'purgecss'
import { createDebug, options } from '../util'
const debug = createDebug('purgecss/purger')

/**
 * Handles purges css files and inline styles
 */
export class Purger {
  /**
   * Cache repository for styles shared between
   * multiple html files
   * @type {Object}
   */
  cachedStyles = {}

  /**
   * Cached contents of script files
   * @type {Object}
   */
  cachedScripts = {}

  /**
   * @type PurgeCSS
   */
  purgeCSS

  /**
   * Options to pass to PurgeCSS
   * @type {Object}
   */
  purgeOptions

  /**
   * @type {AssetMapper}
   */
  mapper

  /**
   * @type {FileWriter}
   */
  writer

  /**
   * Sets PurgeCSS options
   *
   * @constructor
   * @param {AssetMapper} mapper
   * @param {FileWriter} writer
   */
  constructor (mapper, writer) {
    this.purgeCSS = new PurgeCSS()
    this.purgeOptions = { ...options.purgecss }
    if (options.allowSymbols) {
      this.purgeOptions.defaultExtractor = content => content.match(/[\w-/:]+(?<!:)/g) || []
    }
    this.mapper = mapper
    this.writer = writer
    debug('Purger initialized with purgeOptions', this.purgeOptions)
  }

  /**
   * Applies PurgeCSS changes to the given style object
   *
   * @param {Object} style
   * @param {Object} result
   * @param {boolean} fromCache
   * @return {Promise<string[]|null>}
   */
  async applyStyleChanges (style, result, fromCache = false) {
    const rejected = options.purgecss.rejected
      ? result.rejected
      : null
    if (style.type === 'style') {
      style.text.data = result.css
      return rejected
    }

    if (style.type === 'link' && !fromCache) {
      await this.writer.write(style.file, result.css, rejected)
    }
    return []
  }

  /**
   * Purges a style object based on its metadata
   *
   * @param {Object} style
   * @param {HtmlFile} file
   * @return {(Promise<string[]|null>|string[]|null)}
   */
  async purge (style, file) {
    if (style.id && style.id in this.cachedStyles) {
      debug('Retreiving cached purge result for style', style.id)
      return this.applyStyleChanges(style, this.cachedStyles[style.id], true)
    }

    const opts = {
      content: [],
      css: [],
      ...this.purgeOptions
    }
    if (style.type === 'link') {
      try {
        await fs.access(style.file)
        opts.css.push({ raw: await fs.readFile(style.file, 'utf-8') })
      } catch (e) {
        return []
      }
    } else {
      opts.css.push({ raw: style.text.data })
    }

    const files = style.id ? this.mapper.getStyleLinks(style.id) : [file]
    for (const sfile of files) {
      opts.content.push({ raw: sfile.nakedHtml, extension: 'html' })
      await Promise.map(sfile.scripts, async script => {
        if (!(script in this.cachedScripts)) {
          try {
            await fs.access(script)
            this.cachedScripts[script] = await fs.readFile(script, 'utf8')
          } catch (e) {
            return
          }
        }
        opts.content.push({ raw: this.cachedScripts[script], extension: 'js' })
      })
    }

    return this.purgeCSS.purge(opts).then(([result]) => {
      if (style.id) {
        debug('Caching purge result on style', style.id)
        this.cachedStyles[style.id] = result
      }
      return this.applyStyleChanges(style, result)
    })
  }
}
