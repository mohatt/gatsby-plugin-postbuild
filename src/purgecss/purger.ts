import { Promise } from 'bluebird'
import { promises as fs } from 'fs'
import PurgeCSS from 'purgecss'
import { HtmlFile, IStyle, StyleFile } from './html'
import { createDebug, IOptions, options } from '../util'
import { AssetMapper } from './mapper'
const debug = createDebug('purgecss/purger')

export interface PurgeResult {
  css: string
  rejected: string[]
}

/**
 * Handles purges css files and inline styles
 */
export class Purger {
  /**
   * Cache repository for styles shared between
   * multiple html files
   * @type {Object}
   */
  cachedStyles: {
    [id: string]: PurgeResult
  } = {}

  /**
   * Cached contents of script files
   * @type {Object}
   */
  cachedScripts: {
    [id: string]: string
  } = {}

  /**
   * @type PurgeCSS
   */
  purgeCSS

  /**
   * Options to pass to PurgeCSS
   */
  purgeOptions: IOptions['purgecss']

  private readonly mapper

  /**
   * Sets PurgeCSS options
   */
  constructor (mapper: AssetMapper) {
    this.purgeCSS = new PurgeCSS()
    this.purgeOptions = {
      ...options.purgecss,
      rejected: true
    }
    if (options.allowSymbols) {
      this.purgeOptions.defaultExtractor = content => {
        const matchs = content.match(/[\w-/:]+(?<!:)/g)
        if (matchs !== null) return matchs
        return []
      }
    }
    this.mapper = mapper
    debug('Purger initialized with purgeOptions', this.purgeOptions)
  }

  /**
   * Purges a style object based on its metadata
   */
  async purge (style: IStyle, file: HtmlFile): Promise<string[]> {
    if (style.hasID() && style.getID() in this.cachedStyles) {
      if (style instanceof StyleFile) {
        debug('Ignoring rewriting file', style.getID())
        return []
      }
      debug('Retreiving cached purge result for style', style.getID())
      const cache = this.cachedStyles[style.getID()]
      await style.write(cache)
      return cache.rejected
    }

    const opts = this.purgeOptions
    opts.content = []
    opts.css = []
    try {
      opts.css.push({ raw: await style.read() })
    } catch (e) {
      if (style instanceof StyleFile) return []
      throw e
    }

    const files = style.hasID() ? this.mapper.getStyleLinks(style.getID()) : [file]
    for (const sfile of files) {
      opts.content.push({ raw: sfile.nakedHtml, extension: 'html' })
      await Promise.map(sfile.scripts, async script => {
        if (!(script in this.cachedScripts)) {
          try {
            await fs.access(script)
            this.cachedScripts[script] = await fs.readFile(script, 'utf8')
          } catch (e) {
            return
          }
        }
        opts.content.push({ raw: this.cachedScripts[script], extension: 'js' })
      })
    }

    return this.purgeCSS.purge(opts).then(async results => {
      const result = results[0] as PurgeResult
      await style.write(result)
      if (style.hasID()) {
        debug('Caching purge result on style', style.getID())
        this.cachedStyles[style.getID()] = result
      }
      return (style instanceof StyleFile) ? [] : result.rejected
    })
  }
}
