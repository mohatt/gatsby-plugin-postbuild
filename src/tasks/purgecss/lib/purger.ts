import { Promise } from 'bluebird'
import _ from 'lodash'
import PurgeCSS from 'purgecss'
import { createDebug } from '~/common'
import { Filesystem } from '~/filesystem'
import { AssetMapper } from './mapper'
import { HtmlTransformer, HtmlStyle, HtmlStyleFile } from './html'
import { IPurgecssOptions, PurgecssLibOptions, purgecssImportedOptions } from '../options'
const debug = createDebug('purgecss/purger')

export interface IPurgeResult {
  css: string
  rejected: string[]
}

/**
 * Handles purges css files and inline styles
 */
export class Purger {
  /**
   * Cache repository for styles shared between
   * multiple html files
   * @type {Object}
   */
  cachedStyles: {
    [id: string]: IPurgeResult
  } = {}

  /**
   * Cached contents of script files
   * @type {Object}
   */
  cachedScripts: {
    [id: string]: string
  } = {}

  /**
   * PurgeCSS deps
   */
  purgeCSS: PurgeCSS
  purgeOptions: PurgecssLibOptions

  private readonly options: IPurgecssOptions
  private readonly fs: Filesystem
  private readonly mapper

  /**
   * Sets PurgeCSS options
   */
  constructor (options: IPurgecssOptions, fs: Filesystem, mapper: AssetMapper) {
    this.options = options
    this.fs = fs
    this.mapper = mapper
    this.purgeCSS = new PurgeCSS()
    this.purgeOptions = {
      ..._.pick(options, purgecssImportedOptions),
      rejected: true
    }
    if (options.allowSymbols) {
      this.purgeOptions.defaultExtractor = (content: string) => {
        const matchs = content.match(/[\w-/:]+(?<!:)/g)
        if (matchs !== null) return matchs
        return []
      }
    }
    debug('Purger initialized with purgeOptions', this.purgeOptions)
  }

  /**
   * Purges a style object based on its metadata
   */
  async purge (style: HtmlStyle, file: HtmlTransformer): Promise<string[]> {
    if (style.hasID() && style.getID() in this.cachedStyles) {
      if (style instanceof HtmlStyleFile) {
        debug('Ignoring rewriting file', style.getID())
        return []
      }
      debug('Retreiving cached purge result for style', style.getID())
      const cache = this.cachedStyles[style.getID()]
      await style.update(cache)
      return cache.rejected
    }

    const opts = this.purgeOptions
    opts.content = []
    opts.css = []
    try {
      opts.css.push({ raw: await style.read() })
    } catch (e) {
      if (style instanceof HtmlStyleFile) return []
      throw e
    }

    const files = style.hasID() ? this.mapper.getStyleLinks(style.getID()) : [file]
    for (const sfile of files) {
      opts.content.push({ raw: sfile.html, extension: 'html' })
      await Promise.map(sfile.scripts, async script => {
        if (!(script in this.cachedScripts)) {
          try {
            this.cachedScripts[script] = await this.fs.read(script)
          } catch (e) {
            return
          }
        }
        opts.content.push({ raw: this.cachedScripts[script], extension: 'js' })
      })
    }

    return this.purgeCSS.purge(opts).then(async results => {
      const result = results[0] as IPurgeResult
      await style.update(result)
      if (style.hasID()) {
        debug('Caching purge result on style', style.getID())
        this.cachedStyles[style.getID()] = result
      }
      return (style instanceof HtmlStyleFile) ? [] : result.rejected
    })
  }
}
