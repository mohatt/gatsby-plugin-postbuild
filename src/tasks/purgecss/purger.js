import { Promise } from 'bluebird'
import { promises as fs } from 'fs'
import path from 'path'
import PurgeCSS from 'purgecss'
import { options } from '../../util'

export class Purger {
  linkedStyles = {}
  ignoredScripts = []
  cachedScripts = {}
  purgeCSS
  purgeOptions
  writer

  constructor (writer) {
    const ignoredChunks = ['app', 'polyfill']
    const chunks = require(path.join(options._public, 'webpack.stats.json')).assetsByChunkName
    ignoredChunks.forEach(chunk => {
      chunks[chunk].forEach(file => {
        if (this.ignoredScripts.indexOf(file) === -1) {
          this.ignoredScripts.push(file)
        }
      })
    })
    this.purgeCSS = new PurgeCSS()
    this.purgeOptions = { ...options.purgecss }
    if (this.purgeOptions.allowSymbols) {
      delete this.purgeOptions.allowSymbols
      this.purgeOptions.defaultExtractor = content => content.match(/[\w-/:]+(?<!:)/g) || []
    }
    this.writer = writer
  }

  linkStyle (style, file) {
    this.linkedStyles[style.id] ??= { files: [], cache: null }
    if (this.linkedStyles[style.id].files.indexOf(file) === -1) {
      this.linkedStyles[style.id].files.push(file)
    }
  }

  shouldIgnoreScript (script) {
    return this.ignoredScripts.indexOf(script) !== -1
  }

  async applyStyleChanges (style, result, fromCache = false) {
    const rejected = options.purgecss.rejected
      ? result.rejected
      : null
    if (style.type === 'style') {
      style.text.data = result.css
      return rejected
    }

    if (style.type === 'link' && !fromCache) {
      await this.writer.write(style.file, result.css, rejected)
    }
    return []
  }

  async purge (style, file) {
    if (style.id && this.linkedStyles[style.id].cache !== null) {
      return this.applyStyleChanges(style, this.linkedStyles[style.id].cache, true)
    }

    const opts = {
      content: [],
      css: [],
      ...this.purgeOptions
    }

    if (style.type === 'link') {
      try {
        await fs.access(style.file)
        opts.css.push({ raw: await fs.readFile(style.file) })
      } catch (e) {
        return []
      }
    } else {
      opts.css.push({ raw: await style.text.data })
    }

    const files = style.id ? this.linkedStyles[style.id].files : [file]
    for (const sfile of files) {
      opts.content.push({ raw: sfile.nakedHtml, extension: 'html' })
      await Promise.map(sfile.scripts, async script => {
        if (!(script in this.cachedScripts)) {
          try {
            await fs.access(script)
            this.cachedScripts[script] = await fs.readFile(script, 'utf8')
          } catch (e) {
            return
          }
        }
        opts.content.push({ raw: this.cachedScripts[script], extension: 'js' })
      })
    }

    return this.purgeCSS.purge(opts).then(([result]) => {
      if (style.id) {
        this.linkedStyles[style.id].cache = result
      }
      return this.applyStyleChanges(style, result)
    })
  }
}
