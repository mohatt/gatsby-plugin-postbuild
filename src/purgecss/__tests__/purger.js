import { mountFile, mountOptions } from '../../../test/utils'
import { Purger } from '../purger'

jest.mock('fs')
mountOptions()

describe('Purger', () => {
  const writer = { write: jest.fn() }
  const mapper = {}

  it('correctly sets default purgecss extractor', () => {
    mountOptions({
      allowSymbols: true
    })
    let purger
    expect(() => { purger = new Purger(mapper, writer) }).not.toThrow()
    expect(purger.purgeOptions.defaultExtractor('.sel:ec to/r')).toMatchObject(['sel:ec', 'to/r'])
  })

  it('correctly purges styles', async () => {
    mountFile('/foo/bar.css', '.foo{} .bar{width:auto}')
    const fixt = {
      style1: { type: 'style', text: { data: '.foo{} .bar{width:auto}' } },
      style2: { type: 'link', file: '/foo/bar.css' },
      file: { nakedHtml: '<div class="bar"></div>', scripts: [] }
    }
    let purger
    expect(() => { purger = new Purger(mapper, writer) }).not.toThrow()
    await expect(purger.purge(fixt.style1, fixt.file)).resolves.toMatchSnapshot()
    await expect(purger.purge(fixt.style2, fixt.file)).resolves.toMatchSnapshot()
    expect(fixt.style1.text.data).toMatchSnapshot()
    expect(writer.write.mock.calls).toMatchSnapshot()
  })

  it('correctly purges styles against html with scripts', async () => {
    mountFile('/foo/bar.js', 'function func(){ var test="foo" }')
    const fixt = {
      style: { type: 'style', text: { data: '.foo{} .bar{width:auto} .lorem{}' } },
      file: { nakedHtml: '<div class="bar"></div>', scripts: ['/foo/bar.js'] }
    }
    let purger
    expect(() => { purger = new Purger(mapper, writer) }).not.toThrow()
    await expect(purger.purge(fixt.style, fixt.file)).resolves.toMatchSnapshot()
    expect(fixt.style.text.data).toMatchSnapshot()
  })

  it('correctly purges linked styles', async () => {
    const fixt = {
      style: { id: 'foo', type: 'style', text: { data: '.foo{} .bar{width:auto} .lorem{}' } },
      file1: { nakedHtml: '<div class="foo"></div>', scripts: [] },
      file2: { nakedHtml: '<div class="bar"></div>', scripts: [] }
    }
    let purger
    mapper.getStyleLinks = () => [fixt.file1, fixt.file2]
    expect(() => { purger = new Purger(mapper, writer) }).not.toThrow()
    await expect(purger.purge(fixt.style, fixt.file1)).resolves.toMatchSnapshot()
    await expect(purger.purge(fixt.style, fixt.file2)).resolves.toMatchSnapshot()
  })
})
