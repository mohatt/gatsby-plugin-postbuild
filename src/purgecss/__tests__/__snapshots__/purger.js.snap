// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Purger correctly purges linked styles 1`] = `
Array [
  ".lorem",
]
`;

exports[`Purger correctly purges linked styles 2`] = `
Array [
  ".lorem",
]
`;

exports[`Purger correctly purges styles 1`] = `
Array [
  ".foo",
]
`;

exports[`Purger correctly purges styles 2`] = `Array []`;

exports[`Purger correctly purges styles 3`] = `".bar{width:auto}"`;

exports[`Purger correctly purges styles 4`] = `
Array [
  Array [
    "/foo/bar.css",
    ".bar{width:auto}",
    Array [
      ".foo",
    ],
  ],
]
`;

exports[`Purger correctly purges styles against html with scripts 1`] = `
Array [
  ".lorem",
]
`;

exports[`Purger correctly purges styles against html with scripts 2`] = `".foo{} .bar{width:auto}"`;
